### PR TITLE
[TORCH-LINALG] Fix AtenView op to create correct reassociation index

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit fixes AtenView op's Linalg lowering by forming correct
   reassociation index.

Signed-off-by: Abhishek Varma <abhishek@nod-labs.com>

### CODE REVIEW

File: test/Conversion/TorchToLinalg/view.mlir [hunk 1]
```diff
@@ -73,4 +73,21 @@ func.func @torch.aten.view$expandInferredDim(%arg0: !torch.vtensor<[2,6],f32>) -
...
+// CHECK-LABEL: func.func @torch.aten.view$collapseAndThenExpand(
+// CHECK-SAME:              %[[INPUT:.*]]: !torch.vtensor<[64,64],f16>
+// CHECK:         %[[INPUT_TENSOR:.*]] = torch_c.to_builtin_tensor %[[INPUT]] : !torch.vtensor<[64,64],f16> -> tensor<64x64xf16>
+// CHECK:         %[[INPUT_TENSOR_CAST:.*]] = tensor.cast %[[INPUT_TENSOR]] : tensor<64x64xf16> to tensor<64x64xf16>
+// CHECK:         %[[COLLAPSED_TENSOR:.*]] = tensor.collapse_shape %[[INPUT_TENSOR_CAST]] {{\[\[}}0, 1]] : tensor<64x64xf16> into tensor<4096xf16>
+// CHECK:         %[[EXPANDED_TENSOR:.*]] = tensor.expand_shape %[[COLLAPSED_TENSOR]] {{\[\[}}0, 1, 2]] : tensor<4096xf16> into tensor<1x4096x1xf16>
+// CHECK:         %[[CAST_OUTPUT:.*]] = tensor.cast %[[EXPANDED_TENSOR]] : tensor<1x4096x1xf16> to tensor<1x4096x1xf16>
+// CHECK:         %[[OUTPUT:.*]] = torch_c.from_builtin_tensor %[[CAST_OUTPUT]] : tensor<1x4096x1xf16> -> !torch.vtensor<[1,4096,1],f16>
+// CHECK:         return %[[OUTPUT]] : !torch.vtensor<[1,4096,1],f16>
+func.func @torch.aten.view$collapseAndThenExpand(%arg0: !torch.vtensor<[64,64],f16>) -> !torch.vtensor<[1,4096,1],f16> {
```
Comment (ramiro050) [inline test/Conversion/TorchToLinalg/view.mlir:87]: Can you add this as an e2e test instead? See: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

**Review by ramiro050 — CHANGES_REQUESTED** [review]

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/DataMovement.cpp [hunk 2]
```diff
@@ -661,14 +662,19 @@ class ConvertAtenViewOp : public OpConversionPattern<AtenViewOp> {
...
-                               expandedInput.has_value() ? expandedInput.value()
-                                                         : castedInput,
-                               outputAssociations)
-                           .getResult();
+      Value src =
+          expandedInput.has_value() ? expandedInput.value() : castedInput;
+      auto reassociation =
+          getReassociationIndicesForReshape(src.getType(), adjustedResultType);
+      if (!reassociation) {
+          reassociation = outputAssociations;
```
[the remark this answers is not in the document]
Comment (Abhishek-Varma) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:670]: I tried debugging this issue by running couple of experiments and skimming through code snippets as well :-
1. [getReassociationIndicesForReshape](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp#L20-L29) would always switch it's two operands in a way that `operand 1's rank` < `operand 2's rank`.
2. [This part](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp#L66-L67) is bailing out yielding a NULL ReassociationIndex if we use `getReassociationIndicesForReshape` in case of [NativeBatchNorm1DModule_basic](https://github.com/llvm/torch-mlir/blob/main/python/torch_mlir_e2e_test/test_suite/norm_like.py#L118-L140) test case. And `outputAssociations` is correct for this case : `[0] -> [0,1,2]`.
3. `getReassociationIndicesForReshape` outperforms `outputAssociations` computation for the attached [test case](https://github.com/llvm/torch-mlir/blob/fc99d5411f3fdd80384518960fb1007661483e8b/python/torch_mlir_e2e_test/test_suite/reshape_like.py#L127-L142) - as mentioned [here](https://github.com/llvm/torch-mlir/issues/2008#issue-1657463742) in detail.

The current lowering doesn't handle `%2 = torch.aten.view %arg3, %1 : !torch.vtensor<[?],f32>, !torch.list<int> -> !torch.vtensor<[1,?,1],f32>` case quite well.
In this particular case, if `%1` is a list `(1,-1,1)` (i.e. tensor.view(1,-1,1)) the lowering would bail out even before `outputAssociations` / `getReassociationIndicesForReshape` can perform their computations **_<-- This is yet another bug which this PR doesn't aim solve_**.
Whereas if `%1` is a list `(1,0,1)` (i.e. tensor.view(1,0,1)) then `outputAssociations` gives correct result whereas `getReassociationIndicesForReshape` fails (as mentioned in point `2` above).

I even tried refactoring the code to compute `getReassociationIndicesForReshape` prior to computing `outputAssociations` but it seems that there are uses of `outputAssociations` in helping perform computations for the [intermediateShape](https://github.com/llvm/torch-mlir/blob/main/lib/Conversion/TorchToLinalg/DataMovement.cpp#L636-L658) and the refactoring ends up opening another can of worms.

From what I observed, `getReassociationIndicesForReshape` might not be best suited for cases which perhaps need Collapse followed by Expand. And this is something which `outputAssociations` computation performs "better" I believe.

The implementation of the lowering seems to be quite complex, dealing with quite a few edge cases and perhaps can be worked on iteratively.

[the remark this answers is not in the document]
Comment (ramiro050) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:670]: I just ran the e2e test added in this PR on the main branch of torch-mlir and it passes without needing any changes. What error are you seeing when running that test without the changes being added here?

I would also prefer if we don't introduce `getReassociationIndicesForReshape` unless it allows for the removal of parts of the ViewOp pattern. Otherwise it just adds more complexity.

[the remark this answers is not in the document]
Comment (Abhishek-Varma) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:670]: Perhaps I wasn't able to capture the necessary IR via the e2e test case. Let me know what am I missing here.

Just ran through the ToMe model to extract the IR snippet which gets solved only after the changes in this revision are applied and I've isolated that in a small function :-

INPUT IR:
```
func.func @view(%872: !torch.vtensor<[64,64],f16>) -> !torch.vtensor<[1,4096,1],f16>
{
  %int1 = torch.constant.int 1
  %int-1 = torch.constant.int -1
  %873 = torch.prim.ListConstruct %int1, %int-1, %int1 : (!torch.int, !torch.int, !torch.int) -> !torch.list<int>
  %874 = torch.aten.view %872, %873 : !torch.vtensor<[64,64],f16>, !torch.list<int> -> !torch.vtensor<[1,4096,1],f16>
  return %874 : !torch.vtensor<[1,4096,1],f16>
}
```

RUN: `torch-mlir-opt -convert-torch-to-linalg ~/input.mlir`

OUTPUT:
```
error: 'tensor.expand_shape' op expected reassociation map #0 of same rank as expanded memref(3), but got 2
%874 = torch.aten.view %872, %873 : !torch.vtensor<[64,64],f16>, !torch.list<int> -> !torch.vtensor<[1,4096,1],f16>
```

On trying to print what the incorrect IR looks like, I used the `-mlir-print-ir-after-all` and got the following :-
```
// -----// IR Dump After ConvertTorchToLinalg Failed (convert-torch-to-linalg) //----- //
"func.func"() ({
^bb0(%arg0: !torch.vtensor<[64,64],f16>):
  %0 = "torch_c.to_builtin_tensor"(%arg0) : (!torch.vtensor<[64,64],f16>) -> tensor<64x64xf16>
  %1 = "torch.constant.int"() {value = 1 : i64} : () -> !torch.int
  %2 = "torch.constant.int"() {value = -1 : i64} : () -> !torch.int
  %3 = "torch.prim.ListConstruct"(%1, %2, %1) : (!torch.int, !torch.int, !torch.int) -> !torch.list<int>
  %4 = "torch_c.to_i64"(%1) : (!torch.int) -> i64
  %5 = "torch_c.to_i64"(%2) : (!torch.int) -> i64
  %6 = "torch_c.to_i64"(%1) : (!torch.int) -> i64
  %7 = "arith.constant"() {value = 0 : index} : () -> index
  %8 = "arith.constant"() {value = 64 : index} : () -> index
  %9 = "arith.constant"() {value = 1 : index} : () -> index
  %10 = "arith.constant"() {value = 64 : index} : () -> index
  %11 = "tensor.cast"(%0) : (tensor<64x64xf16>) -> tensor<64x64xf16>
  %12 = "tensor.collapse_shape"(%11) {reassociation = [[0, 1]]} : (tensor<64x64xf16>) -> tensor<4096xf16>
  %13 = "tensor.expand_shape"(%12) {reassociation = [[0, 1]]} : (tensor<4096xf16>) -> tensor<1x4096x1xf16>
  %14 = "tensor.cast"(%13) : (tensor<1x4096x1xf16>) -> tensor<1x4096x1xf16>
  %15 = "torch_c.from_builtin_tensor"(%14) : (tensor<1x4096x1xf16>) -> !torch.vtensor<[1,4096,1],f16>
  "func.return"(%15) : (!torch.vtensor<[1,4096,1],f16>) -> ()
}) {function_type = (!torch.vtensor<[64,64],f16>) -> !torch.vtensor<[1,4096,1],f16>, sym_name = "view"} : () -> ()
```

Whereas, if we apply the changes in this patch, we correctly get the following IR :-
```
module {
  func.func @view(%arg0: !torch.vtensor<[64,64],f16>) -> !torch.vtensor<[1,4096,1],f16> {
    %0 = torch_c.to_builtin_tensor %arg0 : !torch.vtensor<[64,64],f16> -> tensor<64x64xf16>
    %int1 = torch.constant.int 1
    %int-1 = torch.constant.int -1
    %1 = torch.prim.ListConstruct %int1, %int-1, %int1 : (!torch.int, !torch.int, !torch.int) -> !torch.list<int>
    %2 = torch_c.to_i64 %int1
    %3 = torch_c.to_i64 %int-1
    %4 = torch_c.to_i64 %int1
    %c0 = arith.constant 0 : index
    %c64 = arith.constant 64 : index
    %c1 = arith.constant 1 : index
    %c64_0 = arith.constant 64 : index
    %cast = tensor.cast %0 : tensor<64x64xf16> to tensor<64x64xf16>
    %collapsed = tensor.collapse_shape %cast [[0, 1]] : tensor<64x64xf16> into tensor<4096xf16>
    %expanded = tensor.expand_shape %collapsed [[0, 1, 2]] : tensor<4096xf16> into tensor<1x4096x1xf16>
    %cast_1 = tensor.cast %expanded : tensor<1x4096x1xf16> to tensor<1x4096x1xf16>
    %5 = torch_c.from_builtin_tensor %cast_1 : tensor<1x4096x1xf16> -> !torch.vtensor<[1,4096,1],f16>
    return %5 : !torch.vtensor<[1,4096,1],f16>
  }
}
```

[the remark this answers is not in the document]
Comment (ramiro050) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:670]: The e2e test should match this example: input of size `(64, 64)`, and output shape `(1, 4096, 1)`. The issue seems to be that when there is collapse and expand, trailing unit-sized dimensions are not correctly handled by the expand.

[the remark this answers is not in the document]
Comment (Abhishek-Varma) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:670]: The e2e test masked the real issue by involving other fixed dimensions.
I made the changes `(64,64) -> (1, 4096, 1)` and see that it errors out saying `mismatching contracting dimension`.
Guess this would involve a deeper look!

Strangely enough the same IR for ToMe model gets solved with this patch.

[the remark this answers is not in the document]
Comment (Abhishek-Varma) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:670]: My bad!

I mistakenly added a pure dynamic tensor shapes, which I see is anyway not supported in the current lowering.

I've added the correct e2e test which would work with this revision's change.

Can you take a look at it now? @ramiro050

[the remark this answers is not in the document]
Comment (ramiro050) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:670]: I still feel that we should not introduce `getReassociationIndicesForReshape` unless it allows for the removal of parts of the `ViewOp` pattern. Patching up bugs without really knowing why the code is not doing the correct things adds complexity that is very hard to simplify later. We should have an understanding of why the current algorithm fails. Once we know what is actually wrong in the logic, we can start looking at how to resolve it.
